firmware/main/main.c

```diff
@@ -0,0 +1,252 @@
+#include "driver/ledc.h"
+#include "esp_err.h"
+#include "esp_log.h"
+
+#include <stdio.h>
+#include <stdlib.h>
+#include <stdint.h>
+#include <math.h>
+
+#include "sdkconfig.h"
+#include "freertos/FreeRTOS.h"
+#include "freertos/task.h"
+
+#include "mpu6050.h"
+
+#include "pin_defs.h"
+#include "utils.h"
+
+#define PI  3.13
+
+// STEP and DIRECTION output pins for stepper motor driver.
+static const gpio_num_t step_pin = GPIO_NUM_16;
+static const gpio_num_t direction_pin = GPIO_NUM_17;
+
+static const uint32_t update_period = 30; // milliseconds to wait between updates
+
+//speed limitations of stepper motor 
+static const int32_t max_freq = 1500;
+static const int32_t min_freq = 20;
+static const int32_t max_freq_change = 10000;
+
+//finds freq
+float find_freq(float freq, float freq_change, int32_t update_period)
+{
+    freq = freq + (freq_change*update_period/1000);
+    return freq;
+}
+
+//sets freq_current
+void set_freq(ledc_mode_t speed_mode, ledc_timer_t timer_num, float freq)
+{
+    if ((int)abs(freq) < min_freq)
+    {
+        gpio_set_level(step_pin, 1);
+    }
+	else if ((int)abs(freq) > max_freq)
+	{
+		ledc_set_freq(speed_mode, timer_num, max_freq);
+	}
+    else
+    {        
+		ledc_set_freq(speed_mode, timer_num, (int)abs(freq));
+    }   
+}
+
+//finds and changes direction 
+void set_direction(float freq)
+{
+    int direction_current = (freq > 0);
+    gpio_set_level(direction_pin, direction_current);
+}
+
+
+
+
+// constants related to system (assumed for now)
+static const float radius = 0.6; // pulley radius in cm
+static const int m_c = 160;     // mass of cart in gm
+static const int m_p = 240;     // maass of pendulum in gm
+
+//initialising required variables
+float freq_change = 0; // Hz. Speed change per update will not exceed this amount
+float freq = 0; 
+float state_setpoint[4] = {0, 0, PI, 0};
```

Comment:
Add definitions for each of the variables used here. And also explain the working of the LQR controller in detail. You can copy the explanation from your notes or Octave script. But it's good to have the definitions in every place they are used for completeness's sake

Reply:
Also write the State space equations and A and B matrices derived from those